curve_apps/parts_connection/driver.py

```diff
@@ -0,0 +1,197 @@
+
+logger = logging.getLogger(__name__)
+
+
+class PartsConnectionDriver(BaseCurveDriver):
+    """
+    Driver for the edge detection application.
+
+    :param parameters: Application parameters.
+    """
+
+    _parameter_class = Parameters
+
+    def __init__(self, parameters: Parameters | InputFile):
+        super().__init__(parameters)
+
+    def run(self):
+        """
+        Driver for Grid2D entity for the automated detection of line features.
+        The application relies on the Canny and Hough transforms from the
+        Scikit-Image library.
+        """
+        with self.workspace.open(mode="r+") as workspace:
+            parent = None
+            if self.params.output.out_group is not None:
+                parent = ContainerGroup.create(
+                    workspace=workspace,
+                    name=self.params.output.ga_group_name,
+                )
+
+            logger.info("Begin connecting labels ...")
+
+            vertices, cells, labels = self.get_connections()
+
+            logger.info("Process completed.")
+
+            if cells is None:
+                logger.info("No connections found.")
+                return
+
+            name = "Parts Connection"
+            if self.params.output.export_as is not None:
+                name = self.params.output.export_as
+
+            edges = Curve.create(
+                workspace=workspace,
+                name=name,
+                vertices=vertices,
+                cells=cells,
+                parent=parent,
+            )
+
+            if edges is not None:
+                if self.params.source.data is not None:
+                    edges.add_data(
+                        {
+                            self.params.source.data.name: {
+                                "values": labels,
+                                "entity_type": self.params.source.data.entity_type,
+                                "association": "VERTEX",
+                            }
+                        }
+                    )
+
+                self.update_monitoring_directory(
+                    parent if parent is not None else edges
+                )
+
+                logger.info(
+                    "Curve object '%s' saved to '%s'.", name, str(workspace.h5file)
+                )
+
+    def get_connections(self) -> tuple:
+        """
+        Find connections between entity parts.
+
+        :params entity: A Curve object with parts.
+        :params data: Input referenced data.
+        :params detection: Detection parameters.
+
+        :returns : n x 3 array. Vertices of connecting lines.
+        :returns : list
+            n x 2 float array. Cells of edges.
+
+        """
+        max_distance = self.params.detection.max_distance
+
+        if max_distance is None:
+            max_distance = np.inf
+
+        path_list = []
+        out_labels = np.zeros_like(self.labels).astype("int32")
+
+        for value in tqdm(np.unique(self.labels)):
```

Comment:
having a title in tqdm to explain the process?

Reply:
I am acutally not sure how to do this. You have an example?

Reply:
tqdm (self, iterable, desc= “Text You want”)